algorithms/src/snark/varuna/varuna.rs

```diff
@@ -66,25 +69,35 @@ impl<E: PairingEngine, FS: AlgebraicSponge<E::Fq, 2>, MM: SNARKMode> VarunaSNARK
         universal_srs: &UniversalSRS<E>,
         circuits: &[&C],
     ) -> Result<Vec<(CircuitProvingKey<E, MM>, CircuitVerifyingKey<E>)>> {
+        ensure!(!circuits.is_empty());
         let index_time = start_timer!(|| "Varuna::CircuitSetup");
 
-        let mut circuit_keys = Vec::with_capacity(circuits.len());
+        let mut indexed_circuits = Vec::with_capacity(circuits.len());
+        let mut degree_info = None::<DegreeInfo>;
         for circuit in circuits {
-            let mut indexed_circuit = AHPForR1CS::<_, MM>::index(*circuit)?;
-            let degree_info = indexed_circuit.index_info.degree_info::<E::Fr, MM>();
-            // TODO: Add check that c is in the correct mode.
-            // Ensure the universal SRS supports the circuit size.
-            universal_srs
-                .download_powers_for(0..degree_info.max_degree)
-                .map_err(|e| anyhow!("Failed to download powers for degree {}: {e}", degree_info.max_degree))?;
+            let indexed_circuit = AHPForR1CS::<_, MM>::index(*circuit)?;
+            let degree_info_i = indexed_circuit.index_info.degree_info::<E::Fr, MM>();
+            degree_info = match degree_info {
+                Some(degree_info) => Some(degree_info.union(&degree_info_i)),
+                None => Some(degree_info_i),
+            };
+            indexed_circuits.push(indexed_circuit);
+        }
+        let degree_info = degree_info.unwrap();
```

Comment:
I'd thought about proposing this change earlier. This subtle change will be a pretty big improvement to key parameter download i/o & memory allocation for powers in the case of multiple circuits. In a future PR We might consider making  lines 75-->86 a separate function so that callers estimate the powers they need to download for their circuit batches in case they need a separate download strategy (i.e. wasm)